Code/Framework/AzCore/AzCore/DOM/DomValue.cpp

```diff
@@ -0,0 +1,1241 @@
+                            {
+                                return;
+                            }
+                        }
+                        result = visitor.EndObject(object.size());
+                    }
+                }
+                else if constexpr (AZStd::is_same_v<Alternative, ArrayPtr>)
+                {
+                    result = visitor.StartArray();
+                    if (result.IsSuccess())
+                    {
+                        const Array::ContainerType& arrayContainer = GetArrayInternal();
+                        for (const Value& entry : arrayContainer)
+                        {
+                            result = entry.Accept(visitor, copyStrings);
+                            if (!result.IsSuccess())
+                            {
+                                return;
+                            }
+                        }
+                        result = visitor.EndArray(arrayContainer.size());
+                    }
+                }
+                else if constexpr (AZStd::is_same_v<Alternative, NodePtr>)
+                {
+                    const Node& node = *AZStd::get<NodePtr>(m_value);
+                    result = visitor.StartNode(node.GetName());
+                    if (result.IsSuccess())
+                    {
+                        const Object::ContainerType& object = GetObjectInternal();
+                        for (const Object::EntryType& entry : object)
+                        {
+                            result = visitor.Key(entry.first);
+                            if (!result.IsSuccess())
+                            {
+                                return;
+                            }
+                            result = entry.second.Accept(visitor, copyStrings);
+                            if (!result.IsSuccess())
+                            {
+                                return;
+                            }
+                        }
+
+                        const Array::ContainerType& arrayContainer = GetArrayInternal();
+                        for (const Value& entry : arrayContainer)
+                        {
+                            result = entry.Accept(visitor, copyStrings);
+                            if (!result.IsSuccess())
+                            {
+                                return;
+                            }
+                        }
+
+                        result = visitor.EndNode(object.size(), arrayContainer.size());
+                    }
+                }
+                else if constexpr (AZStd::is_same_v<Alternative, AZStd::any>)
+                {
+                    result = visitor.OpaqueValue(*arg);
+                }
+            },
+            m_value);
+        return result;
+    }
+
+    AZStd::unique_ptr<Visitor> Value::GetWriteHandler()
+    {
+        return AZStd::make_unique<ValueWriter>(*this);
+    }
+
+    bool Value::DeepCompareIsEqual(const Value& other) const
```

Comment:
It's worth considering adding a function that support equivalency as well, like `DeepCompareIsEquivalent`. The difference for instance is that values that don't match in type can still have an equivalent representation. For instance a bool can also be a number or the string "true" or "false". The Json Serialization supports this at a higher level, which can continue to be the solution, but as the Generic DOM is used in a variety of tools outside the Json Serialization that may benefit from a similar flexibility, it might be an idea to adding to the Generic DOM. It is however a much more complex test, especially because things like localization come into play as well, so it may also be better handled at a different level. Just some food for thought.

Reply:
Hmm, good callout, I'm inclined to see it'll be better off as external API for now due to the level of ambiguity, but I'm pretty open to bringing it in if it ends up sounding useful